### PR TITLE
Feature 4 implement weekly messages

### PR DESCRIPTION
**Pull request by Jdallos**

### CODE REVIEW

File: serverless.ts [hunk 1]
```diff
@@ -2,38 +2,40 @@ import type { AWS } from "@serverless/typescript";
 import * as dotenv from "dotenv"
 dotenv.config({ path: __dirname + "/.env" });
 
-import sendSlack from '@functions/sendslack';
+import sendDailyMessage from "@functions/sendDailyMessage";
+
```
Comment (jarinykanen) [inline serverless.ts]:
```suggestion
```

File: src/features/future/weeklyBreakdown/formatWeeklyBreakdownData.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
+// import { FormattedTimebankData } from "@functions/schema";
```
Comment (jarinykanen) [inline src/features/future/weeklyBreakdown/formatWeeklyBreakdownData.ts]: Remove comments and fix the possible errors. Add comment why this code is not used

File: src/features/future/weeklyBreakdown/handlerFunctionsForBreakdown.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+  //  const lastWeek = DateTime.now().minus({ days: 7 }).toISODate();
```
Comment (jarinykanen) [inline src/features/future/weeklyBreakdown/handlerFunctionsForBreakdown.ts]: Remove comments and fix the possible errors. Add comment why this code is not used

File: src/features/future/weeklyBreakdown/weeklybreakdownmessage.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
+// import { WeeklyFormattedTimebankData } from "@functions/schema";
```
Comment (jarinykanen) [inline src/features/future/weeklyBreakdown/weeklybreakdownmessage.ts]: Remove comments and fix the possible errors. Add comment why this code is not used

File: src/features/generic/time-utils.ts [hunk 5]
```diff
@@ -17,6 +18,21 @@ namespace TimeUtilities {
...
+    export const lastWeekDateProvider = (): Dates => {
+      const weekStartDate = DateTime.now().startOf('week').minus({ weeks: 1 });
+      const weekEndDate = DateTime.now().startOf('week').minus({ days: 1 });
+      const numberedYear = weekEndDate.year;
+      const numberedWeek = weekStartDate.weekNumber;
+      const weekStartString = weekStartDate.toISODate();
+      const weekEndString = weekEndDate.toISODate();
+
+      return { numberedYear, numberedWeek,  weekStartString, weekEndString };
+    }
```
Comment (jarinykanen) [inline src/features/generic/time-utils.ts]:
```suggestion
    /**
     * Generates dates and numbers for previous week
     *
     * @returns various date formats/ numbers for start and end of previous week
     */
    export const lastWeekDateProvider = (): Dates => {
      const startOfWeek = DateTime.now().startOf("week");

      const weekStartDate = startOfWeek.minus({ weeks: 1 });
      const weekEndDate = startOfWeek.minus({ days: 1 });

      return { weekEndDate, weekStartDate };
    }
```

File: src/features/slackapi/slackapi-utils.ts [hunk 6]
```diff
@@ -31,38 +31,94 @@ namespace SlackApiUtilities {
...
    * Create message based on specific users timebank data
    *
-   * @param filteredTimebankData list of relevant timebank data
+   * @param user timebank data
    * @returns string message if id match
    */
-  const constructSingleDayMessage = (filteredTimebankData: FormattedTimebankData) => {
-    const { name, logged, expected, date } = filteredTimebankData;
+  const constructDailyMessage = (user: DailyCombinedData) => {
+    const { name, logged, expected, date, project, internal, difference } = user;
```
Comment (jarinykanen) [inline src/features/slackapi/slackapi-utils.ts]:
```suggestion
    const { name, logged, expected, date, project, internal, difference } = user;

```

File: src/features/slackapi/slackapi-utils.ts [hunk 7]
```diff
@@ -31,38 +31,94 @@ namespace SlackApiUtilities {
...
-          });
+          console.log(constructDailyMessage(user));
+            // client.chat.postMessage({
+            // channel: slackId,
+            // text: constructDailyMessage(user)
+          // });
+        }
+        catch (error) {
+            console.error(`Error while posting slack messages to user ${user.name}`);
         }
```
Comment (jarinykanen) [inline src/features/slackapi/slackapi-utils.ts]:
```suggestion
        client.chat.postMessage({
          channel: slackId,
          text: constructDailyMessage(user)
        });
      } catch (error) {
        console.error(`Error while posting slack messages to user ${user.name}`);
      }
```

File: src/features/slackapi/slackapi-utils.ts [hunk 8]
```diff
@@ -31,38 +31,94 @@ namespace SlackApiUtilities {
...
+   *
+   * @param weeklyCombinedData list of combined timebank and slack user data
+   * @param weekStart when time data starts
+   * @param weekEnd when time data ends
+   */
+  export const postWeeklyMessage = (weeklyCombinedData: WeeklyCombinedData[], weekStart: string, weekEnd: string) => {
+    weeklyCombinedData.forEach(user => {
+      const { slackId } = user;
+
+        try {
```
Comment (jarinykanen) [inline src/features/slackapi/slackapi-utils.ts]: Change to
```
      try {
        client.chat.postMessage({
          channel: slackId,
          text: constructWeeklySummaryMessage(user, weekStart, weekEnd)
        });
      } catch (error) {
        console.error(`Error while posting weekly slack messages to user ${user.name}`);
      }
```

File: src/features/timebank/timebank-API-provider.ts [hunk 9]
```diff
@@ -42,6 +43,29 @@ namespace TimeBankApiProvider {
     }
   };
 
+  /**
+   *
```
Comment (jarinykanen) [inline src/features/timebank/timebank-API-provider.ts:48]: Missing description

File: src/features/timebank/timebank-API-provider.ts [hunk 10]
```diff
@@ -42,6 +43,29 @@ namespace TimeBankApiProvider {
...
+   * @param duration of time data to sum
+   * @param person person data from timebank
+   * @param year of data to request
+   * @param week of data to request
+   * @returns total time data with user name
+   */
+  export const getTotalTimeEntries = async(duration: 'ALL_TIME' | 'YEAR' | 'MONTH' | 'WEEK', person: PersonDto, year: number, week: number): Promise<WeeklyCombinedData[]> => {
```
Comment (jarinykanen) [inline src/features/timebank/timebank-API-provider.ts]: Create enum for duration

File: src/features/timebank/timebank-API-provider.ts [hunk 11]
```diff
@@ -42,6 +43,29 @@ namespace TimeBankApiProvider {
...
+      const selectedWeek = body.filter(timePeriod => timePeriod.id.year === year && timePeriod.id.week === week)[0];
+      const { firstName, lastName } = person;
+      const combinedName = `${firstName} ${lastName}`;
+
+      return [{ selectedWeek, name: combinedName }];
+    } catch (error) {
+      console.error("Error while loading total time entries");
+      Promise.reject(error);
+    }
+  }
```
Comment (jarinykanen) [inline src/features/timebank/timebank-API-provider.ts]:
```suggestion
  export const getTotalTimeEntries = async (duration: Duration, person: PersonDto, year: number, week: number): Promise<WeeklyCombinedData> => {
    try {
      const client = new TimebankApi(process.env.timebank_base_url);

      const { body } = await client.timebankControllerGetTotal(person.id.toString(), duration);
      const selectedWeek = body.filter(timePeriod => timePeriod.id.year === year && timePeriod.id.week === week)[0];

      const { firstName, lastName } = person;
      const combinedName = `${firstName} ${lastName}`;

      return { selectedWeek: selectedWeek, name: combinedName };
    } catch (error) {
      console.error("Error while loading total time entries");
      Promise.reject(error);
    }
  }
```

File: src/features/timebank/timebank-utils.ts [hunk 12]
```diff
@@ -8,33 +8,57 @@ import { PersonDto, TimeEntry } from "src/generated/client/api";
...
 
-      return {
-        id: id,
-        name: combinedName,
-        slackId: slackUser?.id,
-        expected: personsTimeEntries[0].expected,
-        logged: personsTimeEntries[0].logged,
-        date: personsTimeEntries[0].date.toISOString()
-      };
+      if(length === 1){
```
Comment (jarinykanen) [inline src/features/timebank/timebank-utils.ts]:
```suggestion
      if (length === 1){
```

File: src/features/timebank/timebank-utils.ts [hunk 13]
```diff
@@ -8,33 +8,57 @@ import { PersonDto, TimeEntry } from "src/generated/client/api";
...
+  export const combineWeeklyData = (
+    timeData: WeeklyCombinedData[],
+    slackUsers: Member[]
+    ): WeeklyCombinedData[] => (
+      timeData.map(entry => {
+        const slackUser = slackUsers.find(slackUser => slackUser.real_name === entry.name);
+
+        return { ...entry, slackId: slackUser.id };
     })
   );
```
Comment (jarinykanen) [inline src/features/timebank/timebank-utils.ts]:
```suggestion
  export const combineWeeklyData = (timeData: WeeklyCombinedData[], slackUsers: Member[]): WeeklyCombinedData[] => (
    timeData.map(entry => {
      const slackUser = slackUsers.find(slackUser => slackUser.real_name === entry.name);

      return slackUser ? { ...entry, slackId: slackUser.id } : entry;
    })
  );
```

File: src/functions/schema.ts [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Dates interface
+ */
+export interface Dates {
+  numberedYear: number;
+  numberedWeek: number;
+  weekStartString: string;
+  weekEndString: string;
+}
+
```
Comment (jarinykanen) [inline src/functions/schema.ts:53]:
```suggestion
```

File: src/functions/sendWeeklyMessage/handler.ts [hunk 15]
```diff
@@ -0,0 +1,47 @@
+import { ValidatedAPIGatewayProxyEvent, ValidatedEventAPIGatewayProxyEvent } from "@libs/api-gateway";
+import { formatJSONResponse } from "@libs/api-gateway";
+import { middyfy } from "@libs/lambda";
+import TimeUtilities from "src/features/generic/time-utils";
+import SlackApiUtilities from "src/features/slackapi/slackapi-utils";
+import TimeBankApiProvider from "src/features/timebank/timebank-API-provider";
+import TimeBankUtilities from "src/features/timebank/timebank-utils";
+
```
Comment (jarinykanen) [inline src/functions/sendWeeklyMessage/handler.ts]:
```suggestion
```

File: src/functions/sendWeeklyMessage/handler.ts [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+import schema, { WeeklyCombinedData } from "../schema";
+
+/**
+ * main function
```
Comment (jarinykanen) [inline src/functions/sendWeeklyMessage/handler.ts]:
```suggestion
 * Lambda for sending weekly messages
```

File: src/functions/sendWeeklyMessage/handler.ts [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+      message: `Everything went well ${event.body.name}...`,
+      event,
+    });
+  } catch (error) {
+    return formatJSONResponse({
+      message: `Error while sending slack message: ${error}`,
+      event,
+    });
+  }
+};
```
Comment (jarinykanen) [inline src/functions/sendWeeklyMessage/handler.ts]:
```suggestion
 * @param event API Gateway proxy event
 * @returns JSON response
 */
const sendWeeklyMessage: ValidatedEventAPIGatewayProxyEvent<typeof schema> = async (event: ValidatedAPIGatewayProxyEvent<typeof schema>) => {
  try {
    const timebankUsers = await TimeBankApiProvider.getTimebankUsers();
    const slackUsers = await SlackApiUtilities.getSlackUsers();

    const { weekStartDate, weekEndDate } = TimeUtilities.lastWeekDateProvider();
    const timeEntries: WeeklyCombinedData[] = [];

    for (const person of timebankUsers) {
      timeEntries.push(await TimeBankApiProvider.getTotalTimeEntries(Duration.WEEK, person, weekStartDate.year, weekEndDate.weekNumber));
    }

    const weeklyCombinedData = TimeBankUtilities.combineWeeklyData(timeEntries, slackUsers);

    SlackApiUtilities.postWeeklyMessage(weeklyCombinedData, weekStartDate.toISODate(), weekEndDate.toISODate());

    return formatJSONResponse({
      message: `Everything went well ${event.body.name}...`,
      event,
    });
  } catch (error) {
    return formatJSONResponse({
      message: `Error while sending slack message: ${error}`,
      event,
    });
  }
};

```

File: src/functions/index.ts [hunk 18]
```diff
@@ -1 +1,2 @@
-export { default as sendSlack } from './sendslack';
+export { default as sendDailyMessage } from './sendDailyMessage';
+export { default as sendWeeklyMessage } from './sendWeeklyMessage';
```
Comment (jarinykanen) [inline src/functions/index.ts]:
```suggestion
export { default as sendDailyMessage } from "./sendDailyMessage";
export { default as sendWeeklyMessage } from "./sendWeeklyMessage";
```

**Review by jarinykanen — CHANGES_REQUESTED** [review]

File: src/features/future/weeklyBreakdown/formatWeeklyBreakdownData.ts (hunk 2 above)
Comment (Jdallos) [reply src/features/future/weeklyBreakdown/formatWeeklyBreakdownData.ts]: Done

File: src/features/future/weeklyBreakdown/handlerFunctionsForBreakdown.ts (hunk 3 above)
Comment (Jdallos) [reply src/features/future/weeklyBreakdown/handlerFunctionsForBreakdown.ts]: done

File: src/features/future/weeklyBreakdown/weeklybreakdownmessage.ts (hunk 4 above)
Comment (Jdallos) [reply src/features/future/weeklyBreakdown/weeklybreakdownmessage.ts]: done

File: src/features/slackapi/slackapi-utils.ts (hunk 8 above)
Comment (Jdallos) [reply src/features/slackapi/slackapi-utils.ts]: done

File: src/features/timebank/timebank-API-provider.ts (hunk 10 above)
Comment (Jdallos) [reply src/features/timebank/timebank-API-provider.ts]: done

File: src/features/timebank/timebank-API-provider.ts (hunk 9 above)
Comment (Jdallos) [reply src/features/timebank/timebank-API-provider.ts:48]: added

File: src/functions/schema.ts (hunk 14 above)
Comment (Jdallos) [reply src/functions/schema.ts:53]: changed types to DateTime and modify types accordingly when passing to functions

**Review by jarinykanen — APPROVED** [review]

**Pull request merged**